File: go_algorithms/algorithms/RowByColumnEnhanced.go
package algorithms

import (
	"fmt"
	"sync"
	"time"
)

func RowByColumnEnhanced(B, C [][]int) ([][]int, int64) {
	fmt.Println("Método: RowByColumnEnhanced")

	startTime := time.Now().UnixNano()
	var wg sync.WaitGroup
	size := len(C)
	A := make([][]int, size)
	for i := range A {
		A[i] = make([]int, size)
	}
	wg.Add(2)

	go func() {
		defer wg.Done()
		for i := 0; i < size/2; i++ {
			for j := 0; j < size; j++ {
				for k := 0; k < size; k++ {
					A[i][j] += B[i][k] * C[k][j]
				}
			}
		}
	}()

	go func() {
		defer wg.Done()
		for i := size / 2; i < size; i++ {
			for j := 0; j < size; j++ {
				for k := 0; k < size; k++ {
					A[i][j] += B[i][k] * C[k][j]
				}
			}
		}
	}()

	wg.Wait()
	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime
	return A, executionTime
}

File: python_algorithms/algorithms/RowByColumnEnhanced.py
import numpy as np
import threading
import time

def Multiply(B, C):
    print("Método: RowByColumnEnhanced")

    start_time = time.time_ns()
    size = len(C)
    A = np.zeros((size, size), dtype=np.int64)

    def worker(start, end):
        for i in range(start, end):
            for j in range(size):
                for k in range(size):
                    A[i][j] += B[i][k] * C[k][j]

    threads = []
    mid = size // 2

    # Create two threads
    thread1 = threading.Thread(target=worker, args=(0, mid))
    thread2 = threading.Thread(target=worker, args=(mid, size))

    # Start threads
    thread1.start()
    thread2.start()

    # Append threads to the list
    threads.append(thread1)
    threads.append(thread2)

    # Wait for both threads to complete
    for thread in threads:
        thread.join()

    end_time = time.time_ns()
    execution_time = end_time - start_time
    return A, execution_time
